### ISSUE TITLE
Getting exception sending emails with MimeKit, after upgrading from 3.4.2 to 3.4.3

### ISSUE THREAD

**Issue by ajtruckle** [jstedfast/MimeKit#872]
Can anyone sched any light on this exception that was triggered on a users computer when my DLL was used to try and send an email?

Exception:

```
<LogEntry Date="2022-12-23 05:57:59" Severity="Exception" Source="MSAToolsGMailLibrary.MSAToolsGMailLibraryClass.SendEmail" ThreadId="1">
    <Exception Type="System.IO.FileNotFoundException" Source="System.Span`1..ctor">
    <Message>Could not load file or assembly 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.</Message>
         <StackTrace>   at System.Span`1..ctor(T[] array)
                        at MimeKit.Utils.ValueStringBuilder..ctor(Int32 initialCapacity)
                        at MimeKit.Utils.Rfc2047.Encode(FormatOptions options, Encoding charset, String text, Boolean phrase)
                        at MimeKit.Utils.Rfc2047.EncodeText(FormatOptions options, Encoding charset, String text)
                        at MimeKit.Header.EncodeUnstructuredHeader(ParserOptions options, FormatOptions format, Encoding encoding, String field, String value)
                        at MimeKit.Header.EncodeAddressHeader(ParserOptions options, FormatOptions format, Encoding encoding, String field, String value)
                        at MimeKit.Header.FormatRawValue(FormatOptions format, Encoding encoding, String value)
                        at MimeKit.Header.SetValue(FormatOptions format, Encoding encoding, String value)
                        at MimeKit.Header.SetValue(Encoding encoding, String value)
                        at MimeKit.Header..ctor(Encoding encoding, HeaderId id, String value)
                        at MimeKit.HeaderList.set_Item(HeaderId id, String value)
                        at MimeKit.MimeMessage..ctor()
                        at MSAToolsGMailLibrary.MSAToolsGMailLibraryClass.SendEmail(String From, String Subject, String Attachment) 
        </StackTrace>
    </Exception>
</LogEntry>
```

**Comment by jstedfast** [comment]
This is a duplicate of https://github.com/jstedfast/MailKit/issues/1462

The solution in that is to use the following App.config:

```
<dependentAssembly>
    <assemblyIdentity name="System.Runtime.CompilerServices.Unsafe" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
    <bindingRedirect oldVersion="0.0.0.0-6.0.0.0" newVersion="4.0.4.1" />
</dependentAssembly>
```

**Comment by jstedfast** [comment]
Note: The NuGet package version 6.0.0 of System.Runtime.CompilerServices.Unsafe does not equate to the *Assembly Version* which is likely 4.0.4.1 which is why your original App.config does not work. You need to map the *Assembly Version* and not the *NuGet Version*.

**Comment by ajtruckle** [comment]
That does not seem to fix it for me.

But it is late.

Night.
________________________________

**Comment by ajtruckle** [comment]
OK, in my other affected project, it uses System.Memory.4.5.5 which depends on Unsafe.

According to the dependency list it says >= 4.5.3.
________________________________

**Comment by ajtruckle** [comment]
I don't get it. So if 4.5.3 IS 4.0.4.1 ... and I have that in the app.config, why the exception?

I did check both output folders and they were the same time stamps.

**Comment by jstedfast** [comment]
Welcome to DLL Hell? I don't know.

**Comment by ajtruckle** [comment]
I think I might know what is happening. I will try later today and come back with feedback.

**Comment by ajtruckle** [comment]
I am giving up on this for now.

Sub project A has a reference to System.Memory. And it has a binding redirect to use 4.0.1.2 (which is the latest). There is no entry for the Unsafe dependency. It has the latest 6.0.0 downloaded. Admittedly the System.Memory says the dependancy is >= 4.5.3 so this is probably why it works. And, the date of the DLL is 19/02/2020.

Now, sub project B is the one with MimeKit. One issue here is that with 3.4.3 you have:

System.Memory >= 4.5.5
Unsafe >= 6.0.0

I don't see why you have done this. In 3.4.2 there was not even a dependency on Unsafe. In System.Memory needs >= 4.5.3 then MimeKit should be >= 4.5.3. Why is it forcing this conflict? I have to stick with 3.4.2 because it is not breaking the system and I need no binding redirects. Life is simple.

Can't you simply change it to >= 4.5.3? To be consistent?

**Comment by jstedfast** [comment]
If your suggestion would fix the issue, then what explains this? https://github.com/jstedfast/MailKit/issues/1462

In other words, other people were getting the exact same problem you are seeing long before I made the explicit dependency on Unsafe >= 6.0.0. The explicit dependency on 6.0.0 was, in fact, an attempt at fixing this issue or at the very least, making it easier to resolve because if MimeKit depends on the latest, then the compiler *should* bring in 6.0.0 and all of the other libraries should Just Work(tm) if they depend on >= 4.0.4.1 or whatever the version is.

I made the dependency on >= 6.0.0 explicit because other NuGet packages that other people are using have a dependency on Unsafe >= 6.0.0.

If System.Memory was the only NuGet package that depended on Unsafe, then I would not make an explicit dependency and things would likely work as long as no one explicitly added a reference to Unsafe that did not match the exact version used by System.Memory.

Unfortunately, that's not something I can depend on happening.

**Comment by jstedfast** [comment]
BTW, it might help if I knew which TargetFramework you are targeting. Maybe there's an issue with whatever framework you are using but not the others?

**Comment by ajtruckle** [comment]
I am targeting Framework 4.6.2 and don't want to change that.

**Comment by ajtruckle** [comment]
I have only one other package that specifically says dependency of >= 4.5.3 and that is System.Memory. And all of my sub tools are consistent with using 4.6.2 Framework. 

I still don;t se why this is forced. There is no specific reason for >= 6.0.0. After all, the package I use (which is last updated in May this year) is not forcing it. We are not all using 4.8 frameworks for various reasons.

In the linked discussion he stated:

> I have System.Runtime.CompilerServices.Unsafe, Version=6.0.0.0 installed on my system, yet MailKit is referencing System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1 which is not installed.

Yet, if your dependency is 4.0.4.1 OR HIGHER then it should not matter if he had a newer Unsafe should it? That's the real problem to have got fixed somehow IMHO.

I have to find a common ground too against all the nuget packages and until system.memory also uses 6.0.0 then I can't upgrade. And framework 4.6.2 may be a big factor.

**Comment by jstedfast** [comment]
> Yet, if your dependency is 4.0.4.1 OR HIGHER then it should not matter if he had a newer Unsafe should it? That's the real problem to have got fixed somehow IMHO.

It shouldn't, but clearly it does or you wouldn't have the problem you are hitting, either, would you?

That's the problem.

If MimeKit depends on 6.0.0 and System.Memory depends on >= 4.5.3, then how are you hitting this issue?

Likewise, if MimeKit depends on >= 4.5.3 and System.Memory depends on >= 4.5.3 and this other guy's project depends on Unsafe >= 6.0.0, then MimeKit/System.Memory should both Just Work(tm), but they don't.

WHY???

The issue is that unless *everything* depends on the same version of Unsafe, then at least 1 of the assemblies will break.

**Comment by ajtruckle** [comment]
Yeah, I agree. Well, I will just stay with3.4.2 for now as it works. I'll kep an eye on when System.Memory upgrades and I will deal with it then.

**Comment by jstedfast** [comment]
There seems to be a blog explaining things here: https://nickcraver.com/blog/2020/02/11/binding-redirects/

Marc Gravell notes that this is an MSBuild issue in one of the comments in a StackExchange.Redis bug report with the same issue you are hitting in MimeKit: https://github.com/StackExchange/StackExchange.Redis/issues/1128#issuecomment-674059593

**Comment by ajtruckle** [comment]
Yeah. Thanks for that. So:

* Am I supposed to add this "binding redirect to 4.0.4.1" **to all of the 3 projects** that refer to the Unsafe DLL?

I do install the Dlls (like Unsafe) along with my DLL in the respective folders on their PC. On mine I get it to copy them to my development folders. I am using Admin when using VS.

**Comment by jstedfast** [comment]
I think if you add an explicit dependency on Unsafe to your project, then VS will handle creating the app.config file for you.

I think you only need it for the *app* project (and not any of the library projects you may have).

**Comment by ajtruckle** [comment]
But to which?

EmailLib
ToolsLib
OutlookLib

My main app is not using nugets. It is c++ MFC. It just has all the packages in a folder of main app. All use packgage referencing.

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by jstedfast** [comment]
The project that contains your main executable as far as I know.

**Comment by ajtruckle** [comment]
Gotcha, but as mentioned, my master application is a C++ MFC project. It does not use NuGet packages nor have a `app.config` file.

I spent some time re-working my sub projects so that they all end up in their own sub-folders of the main application. This way, their respective DLL files are distinct.

I have a total of four sub projects:

* Google Sync

VB.Net It does not use the Unsafe DLL and is a standalone executable (console app). So it is not affected.

* Outlook Sync

C#.Net It indirectly uses the Unsafe DLL (NuGet 6.0.0). It does so indirectly through the `System.Memory` dependency. It too is a standalone console app. The `app.config` looks like this:

```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
    <configSections>
    </configSections>
    <startup> 
        <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.6.2"/>
    </startup>
  <runtime>
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="System.Net.Http" publicKeyToken="b03f5f7f11d50a3a" culture="neutral"/>
        <bindingRedirect oldVersion="0.0.0.0-4.2.0.0" newVersion="4.1.1.3"/>
      </dependentAssembly>
      <dependentAssembly>
        <assemblyIdentity name="Microsoft.Graph.Core" publicKeyToken="31bf3856ad364e35" culture="neutral"/>
        <bindingRedirect oldVersion="0.0.0.0-1.20.0.0" newVersion="1.20.0.0"/>
      </dependentAssembly>
      <dependentAssembly>
        <assemblyIdentity name="Newtonsoft.Json" publicKeyToken="30ad4fe6b2a6aeed" culture="neutral"/>
        <bindingRedirect oldVersion="0.0.0.0-13.0.0.0" newVersion="13.0.0.0"/>
      </dependentAssembly>
      <dependentAssembly>
        <assemblyIdentity name="System.ValueTuple" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral"/>
        <bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.3.0"/>
      </dependentAssembly>
      <dependentAssembly>
        <assemblyIdentity name="System.Numerics.Vectors" publicKeyToken="b03f5f7f11d50a3a" culture="neutral"/>
        <bindingRedirect oldVersion="0.0.0.0-4.1.4.0" newVersion="4.1.4.0"/>
      </dependentAssembly>
      <dependentAssembly>
        <assemblyIdentity name="System.Buffers" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral"/>
        <bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.3.0"/>
      </dependentAssembly>
	</assemblyBinding>
  </runtime>
</configuration>
```

Date of unsafe DLL is 23/10/2021.

* Tools Dll

C#.Net COM Has NuGet Unsafe 6.0.0 installed. Only needed because of `System.Memory`.

`app.config`:

```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
	<runtime>
		<assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
			<dependentAssembly>
				<assemblyIdentity name="System.ValueTuple" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
				<bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.3.0" />
			</dependentAssembly>
			<dependentAssembly>
				<assemblyIdentity name="System.Memory" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
				<bindingRedirect oldVersion="0.0.0.0-4.0.1.2" newVersion="4.0.1.2" />
			</dependentAssembly>
		</assemblyBinding>
	</runtime>
</configuration>
```

Date of unsafe DLL is 23/10/2021.

* Gmail Tools Dll

C#.Net COM Has NuGet Unsafe 4.5.3 installed. Only needed because of `System.Memory`. Or so I thought. But if I upgrade it to 6.0.0 and run my software, the specific feature still fails, even though MimeKit was still 3.4.2. So it just doesn't like have 6.0.0.

But, the other projects also have `System.Memory` and work fine with Unsafe Nuget 6.0.0. But not this one.

So, I have come to the conclusion that the main culprit here might be:

* Google.Apis.Gmail.v1

![Gmail](https://user-images.githubusercontent.com/7985331/209476501-dc585fda-6d64-40b7-85e3-604d7ab892a3.jpg)

I think it is this tool that forces it. My `PackageReference` for this project:

```
    <PackageReference Include="BouncyCastle">
      <Version>1.8.9</Version>
    </PackageReference>
    <PackageReference Include="Google.Apis.Gmail.v1">
      <Version>1.58.0.2650</Version>
    </PackageReference>
    <PackageReference Include="MimeKit">
      <Version>3.4.2</Version>
    </PackageReference>
    <PackageReference Include="Portable.BouncyCastle">
      <Version>1.9.0</Version>
    </PackageReference>
    <PackageReference Include="System.Runtime.CompilerServices.Unsafe">
      <Version>4.5.3</Version>
    </PackageReference>
```

I am curious. If you try adding `Google.Apis.Gmail.v1` and are targeting framework 4.6.2, do you start to get issues?

And as previously mentioned, using a redirect on my app is not sufficient. I reckon it is somthing else under the hood - maybe deep causing the issue.

So I have to stay with 4.5.3.

**Comment by ajtruckle** [comment]
Nuget 4.5.3 will install 19/02/2020 version.
Nuget 6.0.0 will install 23/10/2021 version.

So there is no 4.5.3 assembly to redirect to?

**Comment by ajtruckle** [comment]
Hang on, what about:

1. Upgrade Mimekit.
2. Downgrade Unsafe back to 4.5.3
3. Add binding redirect 0000 to 6000 to use 4041

I reckon that may work as I would still have 453 dll for unsafe.

**Comment by ajtruckle** [comment]
![Error](https://user-images.githubusercontent.com/7985331/209478870-2ff0d004-6fb1-40d6-9ae3-65cd8c7db1e2.jpg)

If I upgrade MimeKit and Downgrade Unsafe, it won't compile as you can see. Even if I add a binding redirect.

**Comment by ajtruckle** [comment]
Hi! I have reached out to Google team. Tge thing is that the Gmail Nuget package supports framework 4.5 or higher. All others seem to be 4.6.2. Makes sense since 4.6.1 is deprecated. So maybe that is why it is forcing it. 

But in my initial tests with a c# console app using our code it was fine. Odd! 😬 So next is to actually try and do this as c# com dll and call from c++.

**Comment by ajtruckle** [comment]
@jstedfast 
Hi, in a linked discussed they suggest:

> It sounds like you may be in a scenario where assembly binding redirects are infeasible - in which case I suspect you're stuck, basically. The modern .NET ecosystem relies on the ability to use dependencies which depend on different versions of the same package, so long as the version required version is still compatible with the most recent required version. The somewhat-legacy .NET Framework usually handles this with assembly binding redirects; .NET Core and .NET have a slightly different mechanism. But if you're in a situation where those facilities aren't available, you're effectively requiring a single set of transitive dependencies where nothing depends on a different version from the rest of the set. If you can find a set of mutually-compatible versions, that's great - but if you can't, there's nothing library authors can really do to help you.

I do believe this is the case we are in.

**Comment by jstedfast** [comment]
Ouch, yea, sounds like it.

**Comment by joaocib** [comment]
I was facing a back and forth war between System.Rutime.CompilerServices.Unsafe and System.Memory. I have tried many possibilities you may think of. I did manage to solve right now (months later) by simple updating System.Memory to 4.5.5. I'm targeting .net462 and using Costura, just to mention.

**Comment by ajtruckle** [comment]
@joaocib  I never got to the bottom of it. In the end I re-wrote my tool with a brand new .Net8 Console app and everything worked OK.

**Comment by jstedfast** [comment]
FWIW, a while back I added MimeKit.dll.config files to the nuget packages which provide the assembly redirects which are supposedly the work-around to fix this, but obviously if you are still hitting this, then I don't know what the solution could be.

**Comment by DodgeyDan** [comment]
So if we've tried binding redirects up to 6.0 and I get the same thing but the primary components I user and other components rely upon System.Runtime.CompilerServices.Unsafe v6 what other options are there? Im about to tell the client emailing is not possible

**Comment by jplatt-ubishops** [comment]
I know this is an old thread but looks like people are still hitting this issue so I'm adding something I discovered that might shine some light on the issue.

First of all, I am also having the "back and forth war between System.Rutime.CompilerServices.Unsafe and System.Memory" as @joaocib was having. My program was made in .Net 4.8.

My suspicion is that the libraries used by the more recent MailKit version depend on more recent versions of the above so older .net programs can hit this issue.

The interesting bit that may help pinpoint the exact issue is that when I use MailKit in my main program, it works fine. It's only when I try to use it in a child thread (a Task object of System.Threading.Tasks) of my main program does it start complaining about the "Could not load file or assembly". Another thing to note is that where I use MailKit is in a separate class in my program however that doesn't seem to matter because when I call the function directly in my main program, it works. I'm thinking that the issue may be that the required modules are loaded at runtime and accessible by the main program but not child processes.

I think the IntelliSense in VS at one point suggested that I explicitly load the required modules in my child processes. I'm going to try this but I don't want to spend any more time on this (I've already spent almost an entire day trying to debug this) so I will probably update my program to use .net 8 and have done with it. I will report back soon.

**Comment by jstedfast** [comment]
@jplatt-ubishops Try using MimeKit and MailKit from the myget nuget source (daily builds). I have removed the explicit dependency on System.Runtime.CompilerServices.Unsafe.

I'd love to know if that solves this issue (I'll be releasing 4.9.0 with this change in the next few days).

**Comment by jplatt-ubishops** [comment]
@jstedfast I was already using version 4.8.0 in my program (I had the need to send emails from my program recently and had issues with the native System.Net.Mail module so I tried MailKit and ended up here :) ). Are you suggesting I try the "latest build" 4.8.0.1266 ? Note that I have not tried older versions of MailKit either.

**Comment by jplatt-ubishops** [comment]
@jstedfast It is actually complaining about System.Memory in my current configuration and not System.Runtime.CompilerServices.Unsafe:

"Could not load file or assembly 'System.Memory, Version=4.0.1.2, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)"

Although like joaocib, it was complaining about System.Runtime.CompilerServices.Unsafe at one point but I don't remember what configuration I was testing at that time.

Note that the version of System.Memory that MailKit seems to be looking for is 4.0.1.2 whereas I have 4.6.0 in my project hence the assembly version mismatch.

**Comment by jstedfast** [comment]
> Are you suggesting I try the "latest build" 4.8.0.1266?

Yes. The myget.org packages have the most recent GitHub commits (4.8.0.x) whereas nuget.org only has the latest public release (4.8.0.0).

**Comment by jplatt-ubishops** [comment]
OK, I satisfied the dependencies by forcing references to System.Memory v4.0.1.2. Then it wanted System.Buffers v4.0.3.0 and now I'm back to: "Could not load file or assembly 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified."

Going to try the myget latest build 4.8.0.1266 now and see if that fixes it.

**Comment by jplatt-ubishops** [comment]
OK, installed MailKit v4.8.0.1266 and MimeKit v4.8.0.1391.

Now getting: "Could not load file or assembly 'System.Runtime.CompilerServices.Unsafe, Version=6.0.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified."

Which would seem to me is looking for a version from .Net v6 ?

**Comment by jstedfast** [comment]
> Which would seem to me is looking for a version from .Net v6 ?

https://www.nuget.org/packages/System.Runtime.CompilerServices.Unsafe/6.1.0#dependencies-body-tab

Technically, it looks like that package has `net461`, `netstandard2.0`, and `net7.0` assemblies.

> ... bindingRedirects ...

When I dropped the explicit System.Runtime.CompilerServices.Unsafe dependency, I noticed that MSBuild stopped producing a `MimeKit.dll.config` file (containing binding redirects) for the net4x targets.

This might be the issue.

You could try more-or-less cloning the MimeKit.dll.config file from a prior release and just updating the OldVersion/newVersion values. e.g.:

```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <runtime>
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="System.Runtime.CompilerServices.Unsafe" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-6.1.0.0" newVersion="6.1.0.0" />
      </dependentAssembly>
    </assemblyBinding>
  </runtime>
</configuration>
```

**Comment by jplatt-ubishops** [comment]
I did a bunch of housecleaning and uninstalled and reinstalled the packages while cleaning up the app.config files and references in all of my projects.

Currently for binding redirects I have:

`		  <dependentAssembly>
			  <assemblyIdentity name="System.Runtime.CompilerServices.Unsafe" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
			  <bindingRedirect oldVersion="0.0.0.0-6.1.0.0" newVersion="6.1.0.0" />
		  </dependentAssembly>
		  <dependentAssembly>
			  <assemblyIdentity name="System.Memory" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
			  <bindingRedirect oldVersion="0.0.0.0-4.6.0.0" newVersion="4.6.0.0" />
		  </dependentAssembly>
		  <dependentAssembly>
			  <assemblyIdentity name="System.Buffers" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
			  <bindingRedirect oldVersion="0.0.0.0-4.6.0.0" newVersion="4.6.0.0" />
		  </dependentAssembly>
`

I noticed that an app.config was being generated for all the projects in my solution that referenced my project that uses MailKit and contained the same binding redirects. I made sure that all of those had the above binding redirects.

Now I am getting:

"Could not load file or assembly 'System.Buffers, Version=4.0.3.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)"

so I am still at the same dead end and don't know why the binding redirect for System.Buffers seems to be being ignored.

**Comment by ajtruckle** [comment]
Cool. Note that in my case I did correctly have all configuration files present and deployed. But for me this issue was resolved when I rewrote my application in a new .net

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by dundurlunka** [comment]
This is probably one of my first ever comments on a GitHub issue. However, I want to thank you for the 4.9.0 that fixes the problem. Last week I battled with the same problem and yesterday I saw the new version.

My setup is a .NET Framework 4.8. It had MimeKitLite version 4.5.0 which failed when parsing some emails. Upgrading it to 4.9.0 upgraded the notorious System.Runtime.CompilerServices.Unsafe to 6.0.1, System.Buffers to 4.6.0, System.Memory to 4.6.0 and System.Numerics.Vectors to 4.6.0. I let the binding redirects update themselves with the mentioned versions. There was no way of migrating the whole app to .NET 8.

Thanks again, everything works as expected!
